File: src/anchorPreimage.ts
/**
 * anchorPreimage.ts — the exact bytes an external existence-anchor stamps.
 *
 * CURRENT: `payperbyte.io/x402-anchor/receipt/v2-sig` (anchorPreimageV2) — commits the SIGNER,
 * carries the signature alongside, so existence-in-time is checkable without trusting the key and
 * the key is checkable without re-deriving the anchor. The separation mirrors rootcommit/SPEC_SIG.md
 * in MarkovianProtocol/tlog-bitcoin-anchor. SUPERSEDED: `foreseal-receipt-anchor/v1` (anchorPreimage)
 * — the fused digest+sig shape, retained only for lineage (meta.genuine_delivery_anchor_commitment_v1)
 * and the negative-test battery.
 *
 * WHY A FILE FORMAT AND NOT A BARE DIGEST. `ots stamp` takes files, not digests. Hand a stamper a
 * hex digest and the natural move is to write that hex into a file and stamp it — which commits
 * SHA-256(the ASCII hex), not the value you meant. Defining the preimage as a format removes the
 * ambiguity: the stamp commits to the format, not to an ASCII accident.
 *
 * FORMAT (v2-sig) — exactly four LF-terminated lines, no blank lines, no trailing content:
 *
 *   payperbyte.io/x402-anchor/receipt/v2-sig\n
 *   tier=<delivery|provenance>\n
 *   digest=0x<eip712_digest, 32 bytes hex>\n
 *   signer=0x<EIP-55 address the receipt recovers to>\n
 *
 * The commitment is SHA-256 over exactly those bytes. Stamp the file; `ots info <file>.ots` prints
 * that same SHA-256 back, which is the whole digest-binding check and needs no node and no calendar.
 *
 * NO DOMAIN LINE, NO SIGNATURE LINE: the EIP-712 digest already commits to the full domain
 * separator (name, version, chainId, verifyingContract), so a domain line adds no security
 * boundary; and the signature travels alongside the preimage (in the manifest) rather than inside
 * it — that separation is the point of v2-sig.
 */
import { sha256, getAddress, type Hex } from "viem";

export const ANCHOR_PREIMAGE_TAG = "foreseal-receipt-anchor/v1"; // superseded — lineage + negative tests only

/** v2-sig. Tag follows Markovian's `domain/purpose/object/version` convention on a domain we
 *  actually control (they suggested foreseal.io; it does not resolve, and minting a format
 *  identifier on a domain we do not own would age badly). */
export const ANCHOR_PREIMAGE_TAG_V2 = "payperbyte.io/x402-anchor/receipt/v2-sig";

export type Tier = "delivery" | "provenance";

/**
 * v2-sig preimage — four LF-terminated lines:
 *
 *   payperbyte.io/x402-anchor/receipt/v2-sig\n
 *   tier=<delivery|provenance>\n
 *   digest=0x<eip712_digest, 32 bytes hex>\n
 *   signer=0x<EIP-55 address the receipt recovers to>\n
 *
 * The SIGNATURE IS NOT IN THE PREIMAGE. That is the whole point of v2-sig, per Markovian's
 * rootcommit/v2-sig: commit the signER, carry the signATURE alongside. Splitting them means a
 * third party can check existence-in-time without trusting our key, and check our key without
 * re-deriving the anchor. v1 (below) fused `SHA-256(digest ‖ sig)` and is superseded.
 *
 * The receipt's own EIP-712 PayloadAttestation signature is the "who" layer — it recovers to
 * `signer`, so no additional signature is minted. The EIP-712 digest already commits to the full
 * domain separator (name, version, chainId, verifyingContract), so dropping v1's `domain=` line
 * loses no binding — the domain is inside `digest`.
 */
export function anchorPreimageV2(tier: Tier, digest: Hex, signer: string): Uint8Array {
  if (tier !== "delivery" && tier !== "provenance") throw new Error(`unknown tier: ${tier}`);
  const checksummed = getAddress(signer); // EIP-55 form; throws on a malformed address
  // viem's getAddress does NOT reject a bad checksum — it silently re-checksums whatever it is
  // given, so a mistyped signer would be normalised into a valid-looking preimage and stamped,
  // committing an address nobody meant. The EIP-55 checksum exists precisely to catch that typo,
  // so honour it: accept the checksummed form or all-lowercase (which carries no checksum to
  // violate), reject any other casing. Matches conformance/anchor_preimage.py exactly — the
  // cross-impl v2 negative battery pins the two together.
  if (signer !== checksummed && signer !== signer.toLowerCase()) {
    throw new Error(`signer ${signer} has a bad EIP-55 checksum (expected ${checksummed})`);
  }
  const lines = [
    ANCHOR_PREIMAGE_TAG_V2,
    `tier=${tier}`,
    `digest=${requireHex("digest", digest, 32)}`,
    `signer=${checksummed}`,
  ];
  return new TextEncoder().encode(lines.map((l) => l + "\n").join(""));
}

export interface AnchorDomain {
  name: string;
  chainId: number;
}

function requireHex(label: string, v: string, byteLen: number): string {
  const s = v.toLowerCase();
  if (!/^0x[0-9a-f]+$/.test(s) || (s.length - 2) / 2 !== byteLen) {
    throw new Error(`${label} must be 0x-prefixed ${byteLen}-byte hex, got ${v.length} chars`);
  }
  return s;
}

/** The exact stamped bytes. Fails closed on a malformed digest or signature rather than
 *  emitting a preimage that would anchor a value nobody can reproduce. */
export function anchorPreimage(digest: Hex, signature: Hex, domain: AnchorDomain): Uint8Array {
  // The format's entire claim is "exactly four LF-terminated lines" — a newline in the domain name
  // would silently emit five. chainId via BigInt so a large value never renders in exponent form.
  if (/[\r\n]/.test(domain.name)) throw new Error("domain name must not contain CR or LF");
  if (!Number.isInteger(domain.chainId) || domain.chainId < 0) throw new Error("chainId must be a non-negative integer");
  const lines = [
    ANCHOR_PREIMAGE_TAG,
    `domain=eip155:${BigInt(domain.chainId).toString()} ${domain.name}`,
    `digest=${requireHex("digest", digest, 32)}`,
    `sig=${requireHex("sig", signature, 65)}`,
  ];
  return new TextEncoder().encode(lines.map((l) => l + "\n").join(""));
}

/** SHA-256 over the preimage bytes — the value Bitcoin ends up committing to. */
export function anchorCommitment(preimage: Uint8Array): Hex {
  return sha256(preimage);
}

File: conformance/anchor_preimage.py
#!/usr/bin/env python3
"""anchor_preimage.py — Python side of `foreseal-receipt-anchor/v1`.

Independent of the TypeScript emitter (src/anchorPreimage.ts) and must produce BYTE-IDENTICAL
files. Cross-impl agreement is the point: a preimage format only works as an interop shape if two
implementations written separately land on the same bytes.

Format — exactly four LF-terminated lines, no blank lines, no trailing content:

    foreseal-receipt-anchor/v1\n
    domain=eip155:<chainId> <domain name>\n
    digest=0x<eip712_digest, 32 bytes hex>\n
    sig=0x<signature, 65 bytes hex>\n

Commitment = SHA-256 over exactly those bytes. `ots stamp <file>` then commits that value, and
`ots info <file>.ots` prints it back offline.

    python3 conformance/anchor_preimage.py            # emit + verify both tiers, cross-check TS
    python3 conformance/anchor_preimage.py --write DIR # also write the .bin files
"""
from __future__ import annotations

import argparse
import hashlib
import json
import re
from pathlib import Path

TAG = "foreseal-receipt-anchor/v1"
HERE = Path(__file__).resolve().parent
REPO = HERE.parent

def _require_hex(label: str, v: str, byte_len: int) -> str:
    """Fail closed. Uses re.fullmatch with an EXACT nibble count — deliberately not `^...$` with a
    floor-division length test, which accepted two malformed classes the TS emitter rejected:
    an odd nibble count (floor division rounded it away) and a trailing newline (Python's `$`
    matches before a final \\n, which would have emitted a FIVE-line preimage)."""
    s = v.lower()
    if not re.fullmatch(r"0x[0-9a-f]{%d}" % (byte_len * 2), s):
        raise ValueError(f"{label} must be 0x-prefixed {byte_len}-byte hex, got {v!r}")
    return s


def anchor_preimage(digest: str, signature: str, name: str, chain_id: int) -> bytes:
    """The exact stamped bytes. Fails closed on malformed input."""
    if "\n" in name or "\r" in name:
        raise ValueError("domain name must not contain CR or LF")
    if not isinstance(chain_id, int) or chain_id < 0:
        raise ValueError("chainId must be a non-negative integer")
    lines = [
        TAG,
        f"domain=eip155:{chain_id} {name}",
        f"digest={_require_hex('digest', digest, 32)}",
        f"sig={_require_hex('sig', signature, 65)}",
    ]
    return "".join(l + "\n" for l in lines).encode("utf-8")


def anchor_commitment(preimage: bytes) -> str:
    return "0x" + hashlib.sha256(preimage).hexdigest()


def main() -> int:
    ap = argparse.ArgumentParser()
    ap.add_argument("--write", metavar="DIR", help="also write <tier>.preimage.bin into DIR")
    args = ap.parse_args()

    vec = json.loads((HERE / "vector.json").read_text())
    cap = json.loads((REPO / "fixtures" / "live-receipt.json").read_text())
    dom = vec["meta"]["attestation_domain"]
    tiers = {
        "delivery": (vec["meta"]["genuine_delivery_eip712_digest"], cap["attestation"]["signature"]),
        "provenance": (
            vec["meta"]["genuine_provenance_eip712_digest"],
            json.loads(cap["body"])["attestation"]["signature"],
        ),
    }

    rc = 0
    for tier, (digest, sig) in tiers.items():
        pre = anchor_preimage(digest, sig, dom["name"], int(dom["chainId"]))
        commit = anchor_commitment(pre)
        print(f"[{tier}] preimage {len(pre)} B  commitment {commit}")
        print("    " + repr(pre.decode()).replace("\\n", "\\n\n      ")[:0] or "", end="")
        expected = vec["meta"].get(f"genuine_{tier}_anchor_commitment")
        if expected and expected != commit:
            print(f"    MISMATCH vs vector.json: {expected}")
            rc = 1
        if args.write:
            out = Path(args.write) / f"{tier}.preimage.bin"
            out.parent.mkdir(parents=True, exist_ok=True)
            out.write_bytes(pre)
            print(f"    wrote {out} ({len(pre)} B)")
    print("\nPY ANCHOR-PREIMAGE " + ("PASS" if rc == 0 else "FAIL"))
    return rc


if __name__ == "__main__":
    raise SystemExit(main())
